design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,498 @@
+
+```text
+project.yaml
+apis
+├── cluster
+│───├── definition.yaml
+│───├── composition.yaml
+examples
+├── cluster
+├───├── xr.yaml
+functions
+├── compose-cluster
+├───├── main.k
+├───├── helpers.k
+├── propagate-status
+├───├── go.mod
+├───├── go.sum
+├───├── main.go
+├── recycle-nodes
+├───├── main.py
+operations
+├── recycle-nodes
+│───├── operation.yaml
+tests
+├── e2etest-cluster-api
+├───├── test.yaml.gotmpl
+├── test-cluster-api
+├───├── main.py
+├── test-recycle-nodes
+├───├── go.mod
+├───├── go.sum
+├───├── main.go
+```
+
+This example project contains one composite type (`cluster`) supported by two
+functions (`compose-cluster` and `propagate-status`). The `compose-cluster`
+function is built in KCL, while `propagate-status` is built in Go. The project
+also contains one operation (`recycle-nodes`) that runs an operation function of
+the same name, built in Python. The `tests` directory contains tests for the
+composition and operation. We call the functions built as part of a project
+"embedded functions", since they sit alongside configuration rather than in
+their own repositories.
+
+The `project.yaml` file contains metadata and configuration for the project. It
+configures the build tooling and lists the project's dependencies (other
+Crossplane packages such as Providers).
+
+When a user runs `crossplane project build`, four Crossplane packages will be
+produced: a Configuration and three Functions. The Configuration will include
+automatically generated dependencies on the functions, as well as any additional
+dependencies specified in `project.yaml`. Automated tests can be executed with
+`crossplane project test run`, and the project can be installed on a local
+development control plane for manual testing with `crossplane project
+run`. Running `crossplane project push` will push all four packages to a
+registry.
+
+### The Project File
+
+The project configuration file (`project.yaml` by default, overridable with a
+CLI argument) looks like this:
+
+```yaml
+apiVersion: dev.crossplane.io/v1alpha1
+kind: Project
+metadata:
+  name: my-platform
+spec:
+  # These optional fields are converted to Configuration annotations.
+  maintainer: "Platform Team <platform@example.com>"
+  source: github.com/examplecom/my-platform
+  license: Apache-2.0
+  description: An example configuration using functions.
+  readme: This is just an example.
+
+  # OCI repository where the project will be pushed. This is used as part of the
+  # build process to construct dependencies on the embedded functions.
+  repository: ghcr.io/examplecom/my-platform
+
+  # Crossplane version constraints (optional).
+  crossplane:
+    version: ">=v1.17.0-0"
+
+  # External dependencies (optional).
+  dependsOn:
+    - provider: xpkg.crossplane.io/crossplane-contrib/provider-nop
+      version: ">=v0.2.1"
+    - function: xpkg.crossplane.io/crossplane-contrib/function-auto-ready
+      version: ">=v0.2.1"
```

Comment:
this information can also be in crossplane.yaml, why to have them also here (and possibly duplicated)?

Reply:
Perhaps I can clarify this in the text: when using projects, you write only a `project.yaml`, not a `crossplane.yaml`. The `crossplane.yaml` for the resulting package is automatically generated as part of the build step, and the dependencies from the `project.yaml` are copied into it.